File: lib/parser.ml
open Base
module Binding = Binding

let parse_to_yojson_pretty_string tokens_str =
  tokens_str |> Binding.parse |> Yojson.Safe.from_string |> Yojson.Safe.pretty_to_string
;;

let is_error json =
  let open Yojson.Safe.Util in
  match json |> member "error" with
  | `Null -> false
  | _ -> true
;;

let get_error_message json =
  let open Yojson.Safe.Util in
  json |> member "message" |> to_string
;;

let parse tokens_str : (Yojson.Safe.t, string) Result.t =
  let json = tokens_str |> Binding.parse |> Yojson.Safe.from_string in
  if is_error json then Error (get_error_message json) else Ok json
;;

let%test_module "Parser tests" =
  (module struct
    let%expect_test "parse simple" =
      let input =
        {|[[ "1", "WRITE", "wRite" ],
            [ "1", "NUMTOKEN", "1" ],
            [ "1", "POWER", "**" ],
            [ "1", "NUMTOKEN", "1" ],
            [ "1", "SEMICOLON", ";" ]]|}
      in
      parse_to_yojson_pretty_string input |> Stdio.print_endline;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            {
              "type": "WRITE",
              "arg": {
                "type": "POWER",
                "arg": [
                  { "type": "NUMTOKEN", "value": "1" },
                  { "type": "NUMTOKEN", "value": "1" }
                ]
              }
            }
          ]
        }
        |}]
    ;;

    let%expect_test "parse some other stuff" =
      let input =
        {|WRITE "hello";
        WRITE 1 + 1;
      WRITE 1 - 1;
        WRITE 1 * 1;
      |}
      in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            {
              "type": "WRITE",
              "arg": { "type": "STRTOKEN", "value": " \"hello\" " }
            },
            {
              "type": "WRITE",
              "arg": {
                "type": "PLUS",
                "arg": [
                  { "type": "NUMTOKEN", "value": "1" },
                  { "type": "NUMTOKEN", "value": "1" }
                ]
              }
            },
            {
              "type": "WRITE",
              "arg": {
                "type": "MINUS",
                "arg": [
                  { "type": "NUMTOKEN", "value": "1" },
                  { "type": "NUMTOKEN", "value": "1" }
                ]
              }
            },
            {
              "type": "WRITE",
              "arg": {
                "type": "TIMES",
                "arg": [
                  { "type": "NUMTOKEN", "value": "1" },
                  { "type": "NUMTOKEN", "value": "1" }
                ]
              }
            }
          ]
        }
        |}]
    ;;

    let%expect_test "parse null" =
      let input = {|WRITE null;|} in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [ { "type": "WRITE", "arg": { "type": "NULL" } } ]
        }
        |}]
    ;;

    let%expect_test "parse booleans" =
      let input =
        {|WRITE true;
        WRITE false;|}
      in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            { "type": "WRITE", "arg": { "type": "TRUE" } },
            { "type": "WRITE", "arg": { "type": "FALSE" } }
          ]
        }
        |}]
    ;;

    let%expect_test "parse lists" =
      let input =
        {|WRITE [1, 2, 3];
        WRITE ["a", "b"];
        Write [];|}
      in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            {
              "type": "WRITE",
              "arg": {
                "type": "LIST",
                "items": [
                  { "type": "NUMTOKEN", "value": "1" },
                  { "type": "NUMTOKEN", "value": "2" },
                  { "type": "NUMTOKEN", "value": "3" }
                ]
              }
            },
            {
              "type": "WRITE",
              "arg": {
                "type": "LIST",
                "items": [
                  { "type": "STRTOKEN", "value": " \"a\" " },
                  { "type": "STRTOKEN", "value": " \"b\" " }
                ]
              }
            },
            { "type": "WRITE", "arg": { "type": "EMPTYLIST" } }
          ]
        }
        |}]
    ;;

    let%expect_test "parse trace" =
      let input = {|TRACE "foo";|} in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect
        {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            {
              "type": "TRACE",
              "line": "1",
              "arg": { "type": "STRTOKEN", "value": " \"foo\" " }
            }
          ]
        }
        |}]
    ;;

    let%expect_test "with tokenize keyword capitalization" =
      let input =
        {|trace "Hallo" where it is not number;
      trace [10,20,50,100,70,40,55] where it / 2 is within 30 to 60;|}
      in
      let output = Tokenizer.tokenize input in
      let res =
        match output with
        | Ok out -> parse_to_yojson_pretty_string out
        | Error err -> err
      in
      Stdio.print_endline res;
      [%expect {|
        {
          "type": "STATEMENTBLOCK",
          "statements": [
            {
              "type": "TRACE",
              "line": "1",
              "arg": {
                "type": "WHERE",
                "arg": [
                  { "type": "STRTOKEN", "value": " \"Hallo\" " },
                  {
                    "type": "ISNOTNUMBER",
                    "arg": { "type": "VARIABLE", "name": "it", "line": "1" }
                  }
                ]
              }
            },
            {
              "type": "TRACE",
              "line": "2",
              "arg": {
                "type": "WHERE",
                "arg": [
                  {
                    "type": "LIST",
                    "items": [
                      { "type": "NUMTOKEN", "value": "10" },
                      { "type": "NUMTOKEN", "value": "20" },
                      { "type": "NUMTOKEN", "value": "50" },
                      { "type": "NUMTOKEN", "value": "100" },
                      { "type": "NUMTOKEN", "value": "70" },
                      { "type": "NUMTOKEN", "value": "40" },
                      { "type": "NUMTOKEN", "value": "55" }
                    ]
                  },
                  {
                    "type": "ISWITHIN",
                    "arg": [
                      {
                        "type": "DIVIDE",
                        "arg": [
                          { "type": "VARIABLE", "name": "it", "line": "2" },
                          { "type": "NUMTOKEN", "value": "2" }
                        ]
                      },
                      { "type": "NUMTOKEN", "value": "30" },
                      { "type": "NUMTOKEN", "value": "60" }
                    ]
                  }
                ]
              }
            }
          ]
        }
        |}]
    ;;
  end)
;;

File: lark_parser/parser.py
from lark import Lark, Transformer, Tree, v_args

# Grammar definition with proper operator precedence hierarchy
# Precedence (lowest to highest): ampersand, additive, multiplicative, power, atom
grammar = r"""
start: code

code: statementblock

statementblock: statement*

statement: write_stmt
         | assign_stmt
         | trace_stmt
         | if_stmt
         | for_stmt

write_stmt: WRITE expr ";"

assign_stmt: IDENTIFIER ":=" expr ";"

trace_stmt: TRACE expr ";"

if_stmt: IF expr THEN statementblock else_part? ENDIF ";"

else_part: ELSE statementblock

for_stmt: FOR IDENTIFIER IN expr DO statementblock ENDDO ";"

?expr: expr "&" additive -> ampersand_op
     | additive

?additive: additive "+" multiplicative -> plus
         | additive "-" multiplicative -> minus
         | multiplicative

?multiplicative: multiplicative "*" power -> times
               | multiplicative "/" power -> divide
               | power

?power: atom "^" power -> power_op
      | atom

?atom: NUMTOKEN -> num
     | STRTOKEN -> str
     | IDENTIFIER -> var
     | "null" -> null_val
     | "true" -> true_val
     | "false" -> false_val
     | function_call
     | NOW -> now_op
     | CURRENTTIME -> currenttime_op
     | TIMETOKEN -> time_token
     | "[" "]" -> empty_list
     | "[" expr list_rest "]" -> non_empty_list
     | "(" expr ")"

function_call: UPPERCASE expr -> uppercase_op
             | MAXIMUM expr -> maximum_op
             | AVERAGE expr -> average_op
             | INCREASE expr -> increase_op
             | TIME expr -> time_op

list_rest: ("," expr)*

NUMTOKEN: /\d+(\.\d+)?/
STRTOKEN: /"[^"]*"|'[^']*'/
IDENTIFIER: /[a-zA-Z_][a-zA-Z0-9_]*/
WRITE: "write"i
TRACE: "trace"i
IF: "if"i
THEN: "then"i
ELSE: "else"i
ENDIF: "endif"i
FOR: "for"i
IN: "in"i
DO: "do"i
ENDDO: "enddo"i
NOW: "now"i
CURRENTTIME: "currenttime"i
TIME: "time"i
UPPERCASE: "uppercase"i
MAXIMUM: "maximum"i
AVERAGE: "average"i
INCREASE: "increase"i
TIMETOKEN: /\d{1,2}:\d{2}(:\d{2})?/

%import common.WS
%ignore WS
"""


class Transformer(Transformer):
    """Transform parse tree into JSON structure"""

    @v_args(inline=True)
    def var(self, identifier):
        return {
            "type": "VARIABLE",
            "name": str(identifier),
            "line": "0",  # Line info not available in pure Lark parsing
        }

    def null_val(self, items):
        return {"type": "NULL"}

    def true_val(self, items):
        return {"type": "TRUE"}

    def false_val(self, items):
        return {"type": "FALSE"}

    @v_args(inline=True)
    def num(self, token):
        return {"type": "NUMTOKEN", "value": str(token)}

    @v_args(inline=True)
    def str(self, token):
        value = str(token)
        if (value.startswith('"') and value.endswith('"')) or (
            value.startswith("'") and value.endswith("'")
        ):
            value = value[1:-1]
        return {"type": "STRTOKEN", "value": value}

    @v_args(inline=True)
    def time_token(self, token):
        return {"type": "TIMETOKEN", "value": str(token)}

    def _binary_op(self, op_name, items):
        return {"type": op_name, "arg": [items[0], items[1]]}

    @v_args(inline=True)
    def ampersand_op(self, a, b):
        return self._binary_op("AMPERSAND", [a, b])

    @v_args(inline=True)
    def plus(self, a, b):
        return self._binary_op("PLUS", [a, b])

    @v_args(inline=True)
    def minus(self, a, b):
        return self._binary_op("MINUS", [a, b])

    @v_args(inline=True)
    def times(self, a, b):
        return self._binary_op("TIMES", [a, b])

    @v_args(inline=True)
    def divide(self, a, b):
        return self._binary_op("DIVIDE", [a, b])

    @v_args(inline=True)
    def power_op(self, a, b):
        return self._binary_op("POWER", [a, b])

    @v_args(inline=True)
    def uppercase_op(self, token, expr):
        return {"type": "UPPERCASE", "arg": expr}

    @v_args(inline=True)
    def maximum_op(self, token, expr):
        return {"type": "MAXIMUM", "arg": expr}

    @v_args(inline=True)
    def average_op(self, token, expr):
        return {"type": "AVERAGE", "arg": expr}

    @v_args(inline=True)
    def increase_op(self, token, expr):
        return {"type": "INCREASE", "arg": expr}

    def function_call(self, items):
        # function_call dispatches to its sub-rules which return dicts
        return items[0]

    @v_args(inline=True)
    def now_op(self):
        return {"type": "NOW"}

    @v_args(inline=True)
    def currenttime_op(self):
        return {"type": "CURRENTTIME"}

    @v_args(inline=True)
    def time_op(self, token, expr):
        return {"type": "TIME", "arg": expr}

    def empty_list(self, items):
        return {"type": "LIST", "arg": []}

    @v_args(inline=True)
    def non_empty_list(self, first_item, rest_items):
        # Combine first item with rest items
        all_items = [first_item] + rest_items
        return {"type": "LIST", "arg": all_items}

    def list_rest(self, items):
        # list_rest returns items from the ("," ampersand)* rule
        # Each match gives us an ampersand, so items is the list of ampersands
        return items

    @v_args(inline=True)
    def write_stmt(self, write_token, expr):
        # WRITE statement
        return {"type": "WRITE", "arg": expr}

    def trace_stmt(self, items):
        # items contains: [TRACE token, expression dict, ";"]
        # Extract line number from TRACE token and expression from dict
        trace_token = None
        expr = None

        for item in items:
            if isinstance(item, dict):
                expr = item
            elif hasattr(item, "type") and item.type == "TRACE":
                trace_token = item

        # Get line number from TRACE token's position
        line_num = (
            str(trace_token.line)
            if trace_token and hasattr(trace_token, "line")
            else "0"
        )

        return {"type": "TRACE", "line": line_num, "arg": expr}

    @v_args(inline=True)
    def assign_stmt(self, ident, expr):
        # Assignment statement: IDENTIFIER = expr
        return {"type": "ASSIGN", "ident": str(ident), "arg": expr}

    def if_stmt(self, items):
        # items: [IF token, condition_dict, THEN token, thenbranch_dict, [else_part_dict], ENDIF token, semicolon token]
        # Filter out dicts, collect in order
        dicts = []
        for i, item in enumerate(items):
            # print(f"  item {i}: {type(item).__name__} = {item if isinstance(item, dict) else str(item)[:20]}")
            if isinstance(item, dict):
                dicts.append(item)

        condition = dicts[0]
        thenbranch = dicts[1]
        # else_part will be the 3rd dict if it exists
        elsebranch = (
            dicts[2] if len(dicts) > 2 else {"type": "STATEMENTBLOCK", "statements": []}
        )

        return {
            "type": "IF",
            "condition": condition,
            "thenbranch": thenbranch,
            "elsebranch": elsebranch,
        }

    def else_part(self, items):
        # items: [ELSE token, statementblock dict]
        # Return only the dict
        return items[1]

    def for_stmt(self, items):
        # items contains: FOR token, IDENTIFIER token, IN token, expression, DO token, statementblock, ENDDO token, semicolon token
        # Extract IDENTIFIER token and dicts
        varname = None
        dicts_only = []
        for item in items:
            if hasattr(item, "type") and item.type == "IDENTIFIER":
                varname = str(item)
            elif isinstance(item, dict):
                dicts_only.append(item)

        expression = dicts_only[0]
        statements = dicts_only[1]
        return {
            "type": "FOR",
            "varname": varname,
            "expression": expression,
            "statements": statements,
        }

    def statement(self, items):
        # Unwrap the statement from any of the statement types
        return items[0]

    def statementblock(self, items):
        return {"type": "STATEMENTBLOCK", "statements": items}

    def code(self, items):
        # items[0] should be the statementblock dict
        return items[0]

    def start(self, items):
        # start wraps the code
        return items[0]


def parse(input_text):
    """Parse input text and return Python dict"""
    parser = Lark(grammar, parser="lalr", transformer=Transformer())
    try:
        result = parser.parse(input_text)
        return result
    except Exception as e:
        return {"error": True, "message": f"Parse error: {str(e)}"}


if __name__ == "__main__":
    # Test example
    test_input = """
    write 42;
    write "hello";
    write x;
    write 5 + 3;
    write 10 * 2 + 5;
    """

    result = parse(test_input)
    print(result)
